### PR TITLE
Fix various tests to pass when serializers are enabled

### PR DESCRIPTION
**Pull request by benharsh**
This PR implements miscellaneous fixes to get tests working when serializers are enabled (via -suseIOSerializers).

#### Compiler Changes

- fixes a bug in PRIM_GETCID when actual is a ref
- fixes a virtual dispatch bug

#### Module Changes

- Implement serialization for associative distributions
- Fixes compilerWarning line numbers when printing ``_ddata``
- Removes extra spaces when printing in "Map" form with DefaultSerializer
- Improve implementation of reading classes in-place via ``deserialize`` methods

#### Test Changes

- Updates associative userAPI tests to use ``ChplFormat`` instead of an iostyle partially matching ``%ht``.
- Updates .good files for map tests
- Adds some ``serialize``/``deserialize``/``init`` implementations in a few tests

#### Testing

- [x] full local

### CODE REVIEW

File: modules/internal/DefaultAssociative.chpl [hunk 1]
```diff
@@ -944,6 +935,35 @@ module DefaultAssociative {
...
+        for key in dom {
+          fmt.writeKey(f, key);
+          fmt.writeValue(f, arr.dsiAccess(key));
+        }
+      } else {
+        for 0..<dom.dsiNumIndices {
+          const k = fmt.readKey(f, dom.idxType);
+
+          if !dom.dsiMember(k) {
+            // TODO: throw error
```
Comment (mppf) [inline modules/internal/DefaultAssociative.chpl:957]: Did you mean to address this in this PR?

File: modules/internal/OwnedObject.chpl [hunk 2]
```diff
@@ -404,6 +404,12 @@ module OwnedObject {
     }
   }
 
+  @chpldoc.nodoc
+  proc _owned.deserialize(reader, ref deserializer) throws {
+    var tmp = this.chpl_p! : borrowed class;
+    reader.read(tmp);
```
Comment (mppf) [inline modules/internal/OwnedObject.chpl:410]: It looks like this reads into `tmp` and then drops the result on the floor. That doesn't seem right.

File: test/arrays/userAPI/assocArrayAPItest.chpl [hunk 3]
```diff
@@ -16,7 +19,18 @@ proc readArray(ref X) {
   X["ten"]   = 10.0;
 }
 
-proc testAssocArrayAPI(ref X: []) {
+proc testAssocArrayAPI(ref X: [], output:fileWriter(?)=stdout) {
```
Comment (mppf) [inline test/arrays/userAPI/assocArrayAPItest.chpl:22]: Could you rename `output` to `arrayOutput` ? I think that would make it a bit clearer what is going on.

File: modules/internal/DefaultAssociative.chpl [hunk 4]
```diff
@@ -944,6 +935,35 @@ module DefaultAssociative {
     }
   }
 
+  proc chpl_serialReadWriteAssociativeHelper(f, arr, dom) throws
+  where _usingSerializers(f) && !_isDefaultDeser(f) {
```
[the remark this answers is not in the document]
Comment (benharsh) [reply modules/internal/DefaultAssociative.chpl:939]: I'm not exactly sure what you're saying here, but the other overload of ``chpl_serialReadWriteAssociativeHelper`` directly below this one implements the "array form" behavior for default serializer/deserializer. Granted, it does so rather directly and not through the serialization API.

Do you want to see a form that goes through the 'startArray' API? Or are you looking for more testing? Both?

File: modules/internal/DefaultAssociative.chpl (hunk 1 above)
Comment (benharsh) [reply modules/internal/DefaultAssociative.chpl:957]: Not at this time. It's not clear to me what kind of error should be thrown. I can update the comment to reflect that uncertainty.

File: modules/internal/OwnedObject.chpl (hunk 2 above)
Comment (benharsh) [reply modules/internal/OwnedObject.chpl:410]: ``tmp`` is just the inner _owned pointer cast to 'borrowed', so it's still referring to the actual data and the 'read' isn't being lost.

Comment (mppf) [reply modules/internal/OwnedObject.chpl:410]: ah, right, thanks for correcting my understanding

File: modules/internal/DefaultAssociative.chpl (hunk 4 above)
[the remark this answers is not in the document]
Comment (mppf) [reply modules/internal/DefaultAssociative.chpl:939]: Yeah, I was imagining that we need it to go through the startArray API. Nonetheless a TODO comment might suffice.

**Pull request merged**